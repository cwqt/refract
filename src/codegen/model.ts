import { CommentsTypes } from '../public/fields/comments';
import * as Types from '../types';
import { nonNullable } from '../types/utils';
import { alignFields } from './align';
import { block } from './block';
import { column } from './column';

export const model = (model: Types.Blocks.Model): string => {
  const [comments, columns] = extractComments(model.columns);

  return [
    comments,
    block(`model ${model.name}`, alignFields(columns.map(column).join('\n'))),
  ]
    .filter(nonNullable)
    .join('\n')
    .trim();
};

export const extractComments = (
  columns: Types.Column<any>[],
): [outside: string, columns: Types.Column[]] => {
  const isUsualComment = (c: Types.Column<any> | Types.Modifier<any>) => c?.type === CommentsTypes.Comment;
  const isAstComment = (c: Types.Column<any> | Types.Modifier<any>) => c?.type === CommentsTypes.AstComment;
  const isComment = (c: Types.Column<any> | Types.Modifier<any>) => isUsualComment(c) || isAstComment(c);

  return [
    // All comment rows for a model are placed outside the model block def
    columns
      .filter(c => isComment(c))
      .map(c => isAstComment(c) ? `/// ${c.modifiers[0].value}` : `// ${c.modifiers[0].value}`)
      .join('\n'),

    columns
      // Remove Comment and AstComment rows to prevent re-insertion
      .filter(c => !isComment(c))
      // Shift all comment modifiers to be on their own row as a Comment column
      .reduce(
        (cols, column) => [
          ...cols,
          ...column.modifiers
            .filter(c => isComment(c))
            .map(c => ({
              name: 'comment',
              type: c.type,
              modifiers: [c],
            })),
          {
            ...column,
            modifiers: column.modifiers.filter(c => !isComment(c)),
          },
        ],
        [],
      ),
  ];
};
