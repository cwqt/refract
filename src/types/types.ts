import { JsonValue } from '../codegen/lib/json';
import { Model } from './blocks';
import { ReferentialAction } from './fields';

type Append<T, K> = { [index in keyof T]: T[index] & K };

type Scalars = Append<
  {
    Int: {
      unique?: true;
      id?: true;
      default?: 'cuid()' | 'autoincrement()' | 'uuid()' | number;
    };
    Float: {
      unique?: true;
      default?: number;
    };
    BigInt: {
      unique?: true;
      default?: BigInt;
    };
    Bytes: {
      unique?: true;
      default?: never; // FIXME: what should this be?
    };
    Decimal: {
      unique?: true;
      default?: number;
    };
    String: {
      unique?: true;
      id?: true;
      default?: string | 'auto()';
      limit?: number;
    };
    Boolean: {
      unique?: true;
      default?: boolean;
    };
    DateTime: {
      default?: 'now()';
      updatedAt?: true;
    };
    Json: { default?: JsonValue };
    // TODO: https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#unsupported
    // Unsupported: {},
  },
  { nullable?: true; map?: string; ignore?: true; raw?: string }
>;

type Enums = {
  Enum: {
    id?: true;
    nullable?: true;
    default?: string;
    ignore?: true;

    // Enum of which this is from
    enum: string;
  };
  // An entry in the enum, e.g. Enum("name", Key("Id", Map("_id")))
  EnumKey: {
    map?: string;
  };
};

type Relations = Append<
  {
    OneToMany: {};
    OneToOne: {
      fields?: string[];
      references?: string[];
      onUpdate?: ReferentialAction;
      onDelete?: ReferentialAction;
      nullable?: true;
    };
    ManyToOne: {
      fields: string[];
      references: string[];
      onUpdate?: ReferentialAction;
      onDelete?: ReferentialAction;
      nullable?: true;
    };
  },
  { name?: string; model: Model }
>;

export type TypeData = Scalars &
  Enums &
  Relations & {
    // Escape hatch -----------------------------------------
    Raw: { value: string };
  };

// All possible column datatypes & their accepted modifiers/parameters
export type Type = keyof TypeData;
