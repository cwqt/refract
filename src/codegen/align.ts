// Align contents of a datasource/generator to that of the longest column
export const alignKv = (value: string): string => {
  const lines = value.split('\n');
  const delimiter = '=';

  const max =
    Math.max(...lines.map(l => l.trim().split(delimiter).shift().length)) + 1;

  return lines
    .map(line => line.split(delimiter))
    .map(([head, ...rest]) =>
      [head.padEnd(max), delimiter, ...rest.map(v => v.trim())].join(' '),
    )
    .join('\n');
};

// Align contents of a model to that of the longest column
export const alignFields = (value: string): string => {
  const lines = value.split('\n');

  let maximumColumnName = 0,
    MaximumColumnType = 0;
  for (const line of lines) {
    const [columnName, columnType] = line.split(' ');
    maximumColumnName = Math.max(maximumColumnName, columnName.length);
    MaximumColumnType = Math.max(MaximumColumnType, (columnType ?? '').length);
  }

  return lines
    .map(line => line.split(' '))
    .map(([columnName, columnType, ...rest]) =>
      [
        columnName.padEnd(maximumColumnName + 1),
        columnType ? columnType.padEnd(MaximumColumnType + 1) : '',
        ...rest.map(v => v.trim()),
      ].join(' '),
    )
    .join('\n');
};
