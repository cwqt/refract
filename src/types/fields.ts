import { Column } from './columns';
import { Modifier, Modifiers } from './modifiers';
import { FieldType, TypeData } from './types';
import { UnionToIntersection } from './utils';
import * as Types from './types';
import { Provider } from './config';
import { DbModifier } from '../public/db/utils';

// Column data-type, String, Int etc.
export type Field<
  T extends FieldType,
  M extends Modifiers<T> = Modifiers<T>,
> = {
  type: T;
  modifiers: Modifier<T, M>[];
};

export type EnumKey<T extends string = string> = {
  name: T;
  modifiers: Modifier<'EnumKey'>[];
};

export type Scalar = keyof Types.Scalars;
export type Enum = keyof Types.Enums;
export type Type = keyof Types.Type;
export type Relation = keyof Types.Relations;
export type Compound = keyof Types.Compounds;

export type Any =
  | Scalar
  | Relation
  | Enum
  | Type
  | Compound
  | 'Raw'
  | 'Unsupported';

export type ReferentialAction =
  | 'Cascade'
  | 'Restrict'
  | 'NoAction'
  | 'SetNull'
  | 'SetDefault';

// Top type for columns
type TopColumn = {
  name: string;
  type: FieldType;
  modifiers: Array<{
    type: keyof UnionToIntersection<
      { [type in FieldType]: TypeData[type] }[FieldType]
    >;
    value: any;
  }>;
};

export function isRaw(column: TopColumn): column is Column<'Raw'> {
  return column.type == 'Raw';
}

export function isComment(column: TopColumn): column is Column<'Comment'> {
  return column.type == 'Comment';
}

export function isCompound(column: TopColumn): column is Column<Compound> {
  return column.type.startsWith('@@');
}

export function isEnumKey(column: TopColumn): column is Column<'EnumKey'> {
  return column.type == 'EnumKey';
}

export function isEnum(column: TopColumn): column is Column<'Enum'> {
  return column.type == 'Enum';
}

export function isType(column: TopColumn): column is Column<'Type'> {
  return column.type == 'Type';
}

export function isUnsupported(
  column: TopColumn,
): column is Column<'Unsupported'> {
  return column.type == 'Unsupported';
}

export function isRelation(column: TopColumn): column is Column<Relation> {
  return ['OneToMany', 'ManyToOne', 'OneToOne'].includes(column.type);
}

export function isDbModifier(
  column: Modifier<any>,
): column is DbModifier<string, Provider, string, any> {
  return (column.type as string).startsWith('@db');
}

export function isScalar(column: TopColumn): column is Column<Scalar> {
  return (
    typeof column == 'object' &&
    [
      isRelation(column),
      isEnumKey(column),
      isEnum(column),
      isCompound(column),
    ].every(v => v == false)
  );
}
