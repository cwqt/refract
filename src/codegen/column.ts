import { modifier } from './modifiers';
import * as Types from '../types';
import { isArray, isString, nonNullable } from '../types/utils';

// Converts a Column to a Prisma row string
export const column = (column: Types.Column): string => {
  if (Types.Fields.isUnsupported(column)) return unsupported(column);
  if (Types.Fields.isCompound(column)) return compound(column);
  if (Types.Fields.isComment(column))
    return `\t// ${column.modifiers[0].value}`;
  if (Types.Fields.isRaw(column)) return `\t${column.modifiers[0].value}`;
  if (Types.Fields.isEnum(column)) return enumeration(column);
  if (Types.Fields.isType(column)) return typecolumn(column);
  if (Types.Fields.isEnumKey(column)) return enumKey(column);
  if (Types.Fields.isScalar(column)) return scalar(column);
  if (Types.Fields.isRelation(column)) return relationship(column);

  throw new Error(
    `CodegenError: Couldn't figure out type for column: ${column.name}`,
  );
};

const unsupported = (column: Types.Column<'Unsupported'>): string => {
  const isNullable = column.modifiers.find(({ type }) => type == 'nullable');

  return `\t${column.name} Unsupported${modifier(
    column.type,
    column.modifiers[0],
  )}${isNullable ? '?' : ''}`;
};

// enum { Foo Bar }
const enumKey = (column: Types.Column<'EnumKey'>) =>
  `\t${column.name} ${column.modifiers
    .map(m => modifier<'EnumKey'>(column.type, m))
    .join(' ')}`.trimEnd();

export const enumeration = (column: Types.Column<'Enum'>) => {
  const [type, ...modifiers] = column.modifiers;
  const isNullable = modifiers.find(({ type }) => type == 'nullable');

  return `\t${column.name} ${type.value}${isNullable ? '?' : ''} ${modifiers
    .map(m => modifier(column.type, m))
    .join(' ')}`.trimEnd();
};

export const typecolumn = (column: Types.Column<'Type'>) => {
  const [type, ...modifiers] = column.modifiers;
  const isNullable = modifiers.find(({ type }) => type == 'nullable');

  return `\t${column.name} ${type.value}${isNullable ? '?' : ''} ${modifiers
    .map(m => modifier(column.type, m))
    .join(' ')}`.trimEnd();
};

const scalar = (column: Types.Column<Types.Fields.Scalar>) => {
  const isNullable = column.modifiers.find(({ type }) => type == 'nullable');
  const isArray = column.modifiers.find(({ type }) => type == 'array');

  return `\t${column.name} ${column.type}${isArray ? '[]' : ''}${
    isNullable ? '?' : ''
  } ${column.modifiers.map(m => modifier(column.type, m)).join(' ')}`.trimEnd();
};

const compound = (column: Types.Column<Types.Fields.Compound>) => {
  if (column.type == '@@ignore') return `\t${column.type}`;
  if (column.type == '@@map')
    return `\t${column.type}(${`"${column.modifiers[0].value}"`})`;

  const map = column.modifiers.find(v => (v.type as 'values' | 'map') == 'map');
  const args = [
    `[${(column.modifiers[0].value as string[]).join(', ')}]`,
    map ? `map: "${map.value}"` : null,
  ].filter(nonNullable);

  return `\t${column.type}(${args.join(', ')})`;
};

const relationship = (column: Types.Column<Types.Fields.Relation>) => {
  if (column.type == 'OneToOne' || column.type == 'ManyToOne') {
    const modifiers = column.modifiers as Types.Modifier<
      'OneToOne' | 'ManyToOne'
    >[];
    const isNullable = modifiers.find(({ type }) => type === 'nullable');

    const [model, ...restModifiers] = modifiers.filter(
      ({ type }) => type !== 'nullable',
    ) as [
      Types.Modifier<'OneToOne' | 'ManyToOne', 'model'>,
      ...Types.Modifier<
        'OneToOne' | 'ManyToOne',
        'name' | 'fields' | 'references' | 'onUpdate' | 'onDelete'
      >[],
    ];

    const relationModifier = restModifiers.length
      ? `@relation(${restModifiers
          .sort(({ type }) => (type === 'name' ? -1 : 0))
          .map(({ type, value }) =>
            type === 'name'
              ? `"${value}"`
              : `${type}: ${isArray(value) ? `[${value.join(', ')}]` : value}`,
          )
          .join(', ')})`
      : '';

    return `\t${column.name} ${
      isString(model.value) ? model.value : model.value.name
    }${isNullable ? '?' : ''} ${relationModifier}`.trimEnd();
  }

  if (column.type == 'OneToMany') {
    const [model, relationName] = column.modifiers as unknown as [
      Types.Modifier<'OneToMany', 'model'>,
      Types.Modifier<'OneToMany', 'name'>,
    ];

    const relationModifier = relationName
      ? `@relation("${relationName.value}")`
      : '';

    return `\t${column.name} ${
      isString(model.value) ? model.value : model.value.name
    }[] ${relationModifier}`.trimEnd();
  }
};
