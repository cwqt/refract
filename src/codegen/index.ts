import { performance } from 'perf_hooks';
import * as Types from '../types';
import { validate } from '../types';
import { del, nonNullable } from '../types/utils';
import { alignKv } from './align';
import { block, header } from './block';
import { enumeration } from './enum';
import { dedent } from './lib/dedent';
import { pipe } from './lib/pipe';
import { model } from './model';
import { type } from './type';
import { kv } from './transform';
import { validateModel } from './validate';

// Takes a Config input & returns a generated Prisma schema file as a string
// which can then be written to a file / formatted by Prisma CLI
export default (
  config: Types.Config,
): { schema: string; time: number; output: string } => {
  const start = performance.now();

  // Allow direct imports, e.g. `import * as schema from './foo'`
  const schema = Array.isArray(config.schema)
    ? config.schema
    : Object.values(config.schema);

  config = validate({ ...config, schema });

  const datasource = config.datasource;
  const generators = config.generators;
  const enums = schema.filter(Types.Blocks.isEnum);
  const types = schema.filter(Types.Blocks.isType);
  const models = schema.filter(Types.Blocks.isModel);

  const group = (header: string, blocks: string[]): string | null =>
    blocks.length == 0 ? null : [header, blocks.join('\n\n')].join('\n\n');

  const generated = dedent(
    [
      header('datasource'),
      block('datasource db', alignKv(kv(datasource))),
      group(
        header('generators'),
        generators.map(generator =>
          block(
            `generator ${generator.name}`,
            alignKv(kv(del(generator, 'name'))),
          ),
        ),
      ),
      group(header('enums'), enums.map(enumeration)),
      group(header('types'), types.map(type)),
      group(header('models'), models.map(pipe(validateModel(config), model))),
    ]
      .filter(nonNullable)
      .flat()
      .join('\n\n'),
  );

  const end = performance.now();
  const time = Number((end - start).toFixed(3));

  return {
    time,
    output: config.output,
    schema: [
      header(
        `refract https://github.com/cwqt/refract - generated in ${time} ms`,
      ),
      generated,
    ].join('\n'),
  };
};
