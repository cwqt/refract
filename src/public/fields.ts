import * as Types from '../types';
import { nonNullable } from '../types/utils';
import { Enum as CallableEnum } from './enum';
import { ReferentialAction, Relation } from '../types/fields';

// Scalars ------------------------------------------------
const scalar =
  <T extends Types.Type>(type: T) =>
  <M extends Types.Modifiers<T>>(...modifiers: Types.Modifier<T, M>[]) => ({
    type,
    modifiers,
  });

export const Int = scalar('Int');
export const String = scalar('String');
export const Float = scalar('Float');
export const BigInt = scalar('BigInt');
export const Bytes = scalar('Bytes');
export const Boolean = scalar('Boolean');
export const Json = scalar('Json');
export const DateTime = scalar('DateTime');
export const Decimal = scalar('Decimal');

// Enums --------------------------------------------------
export const Enum = <E extends Types.Fields.EnumKey[]>(
  name: string,
  ...keys: E
): (<M extends Types.Modifiers<'Enum'>>(
  initial?: E[number]['name'] | null,
  ...modifiers: Types.Modifier<'Enum', M>[]
) => Types.Fields.Field<'Enum', M>) &
  Types.Blocks.Enum => new CallableEnum(name, keys) as any; // _call

export const Key = <T extends string>(
  name: T,
  ...modifiers: Types.Modifier<'EnumKey'>[]
): Types.Fields.EnumKey<T> => ({ name, modifiers });

// Relationships ------------------------------------------
export const OneToMany = <M extends Types.Blocks.Model>(
  model: M,
  ...modifiers: Types.Modifier<'OneToMany'>[]
) =>
  ({
    type: 'OneToMany',
    modifiers: [{ type: 'model', value: model }, ...modifiers],
  } as Types.Fields.Field<'OneToMany'>);

export const RelationName = <T extends Relation>(
  name: string,
): Types.Modifier<T, 'name'> => ({
  type: 'name',
  value: name,
});

export const Fields = <T extends 'OneToOne' | 'ManyToOne'>(
  ...fields: string[]
): Types.Modifier<T, 'fields'> => ({
  type: 'fields',
  value: fields,
});

export const References = <T extends 'OneToOne' | 'ManyToOne'>(
  ...references: string[]
): Types.Modifier<T, 'references'> => ({
  type: 'references',
  value: references,
});

export const OnUpdate = <T extends 'OneToOne' | 'ManyToOne'>(
  referentialAction: ReferentialAction,
): Types.Modifier<T, 'onUpdate'> => ({
  type: 'onUpdate',
  value: referentialAction,
});

export const OnDelete = <T extends 'OneToOne' | 'ManyToOne'>(
  referentialAction: ReferentialAction,
): Types.Modifier<T, 'onDelete'> => ({
  type: 'onDelete',
  value: referentialAction,
});

export const ManyToOne = <M extends Types.Blocks.Model>(
  model: M | string,
  ...modifiers: Types.Modifier<'ManyToOne'>[]
) => {
  return {
    type: 'ManyToOne' as const,
    modifiers: [
      {
        type: 'model',
        value: model,
      },
      ...modifiers,
    ].filter(nonNullable),
  } as Types.Fields.Field<'ManyToOne'>;
};

export const OneToOne = <M extends Types.Blocks.Model>(
  model: M,
  ...modifiers: Types.Modifier<'OneToOne'>[]
) => {
  return {
    type: 'OneToOne' as const,
    modifiers: [
      {
        type: 'model',
        value: model,
      },
      ...modifiers,
    ].filter(nonNullable),
  } as Types.Fields.Field<'OneToOne'>;
};
