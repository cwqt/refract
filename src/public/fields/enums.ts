import * as Types from '../../types';
import { isString, nonNullable } from '../../types/utils';
import { CommentType, Comment, CommentsTypes } from './comments';

export const Key = <T extends string>(
  ...args:
    | [name: T, ...modifiers: Types.Modifier<'EnumKey'>[]]
    | [name: T, ...modifiers: Types.Modifier<'EnumKey'>[], comment: CommentType | string]
): Types.Fields.EnumKey<T> => {
  const [name, ...modifiers] = args;

  let lastElement = args[args.length - 1];

  if (isString(lastElement)) {
    lastElement = Comment(lastElement);
  }

  return lastElement.type === CommentsTypes.Comment || lastElement.type === CommentsTypes.AstComment
    ? {
        name,
        modifiers: [
          ...(modifiers.slice(
            0,
            modifiers.length - 1,
          ) as unknown as Types.Modifier<'EnumKey'>[]),
          { type: lastElement.type as unknown as 'comment', value: lastElement.value as string },
        ],
      }
    : { name, modifiers: modifiers as Types.Modifier<'EnumKey'>[] };
};

class $Enum<K extends Types.Fields.EnumKey[]>
  extends Function
  implements Types.Blocks.Block<'enum'>
{
  type: 'enum' = 'enum' as const;
  columns: Types.Column<'EnumKey' | 'Comment'>[];

  constructor(public name: string, comment: CommentType | string | null, keys: K) {
    super();

    // Define the keys of the enum
    this.columns = keys.map(
      k =>
        ({
          name: k.name,
          type: 'EnumKey' as const,
          modifiers: k.modifiers,
        } as Types.Column<'EnumKey' | 'Comment'>),
    );

    if (comment) {
      if (isString(comment)) {
        comment = Comment(comment);
      }
      this.columns.unshift({
        name: 'comment',
        type: comment.type,
        modifiers: [{ type: 'value', value: comment.value }],
      } as unknown as Types.Column<'EnumKey' | 'Comment'>);
    }

    // evil object reference proxy hacking _call overrides
    // gives us nice curried classes
    return new Proxy(this, {
      apply: (
        target,
        _,
        args: [K[number] | null, ...Types.Modifier<'Enum'>[]],
      ) =>
        target._call(
          args[0] as K[number],
          ...(args.slice(1, args.length - 1) as Types.Modifier<'Enum'>[]),
        ),
    });
  }

  _call(
    initial: K[number] | null = null,
    ...modifiers: Types.Modifier<'Enum'>[]
  ): Types.Fields.Field<'Enum'> {
    return {
      type: 'Enum' as const,
      modifiers: [
        // welcome to hell
        { type: 'enum' as const, value: this.name },
        initial
          ? { type: 'default' as const, value: initial }
          : initial === null
          ? { type: 'nullable' as const, value: true }
          : null,
        ...modifiers,
      ].filter(
        (v, i, a) => nonNullable(v) && a.findIndex(m => m.type == v.type) === i,
      ) as Types.Modifier<'Enum'>[],
    };
  }
}

// _call signature
type R<E extends Types.Fields.EnumKey[]> = (<M extends Types.Modifiers<'Enum'>>(
  initial?: E[number]['name'] | null,
  ...modifiers: Types.Modifier<'Enum', M>[]
) => Types.Fields.Field<'Enum', M>) &
  Types.Blocks.Enum;

export function Enum<E extends Types.Fields.EnumKey[]>(
  name: string,
  ...keys: E
): R<E>;
export function Enum<E extends Types.Fields.EnumKey[]>(
  name: string,
  comment: CommentType | string,
  ...keys: E
): R<E>;
export function Enum<E extends Types.Fields.EnumKey[]>(
  name: string,
  ...args: [CommentType | string, ...E] | E
): R<E> {
  let [comment, ...keys] = args;
  
  if (typeof comment === 'string') {
    comment = Comment(comment);
  }
  
  return new $Enum(
    name,
    comment ? comment as CommentType : null,
    (keys as E),
  ) as any; // _call
}
