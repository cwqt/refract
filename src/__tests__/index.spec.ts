import {
  Array,
  Default,
  Fields,
  Id,
  Int,
  ManyToOne,
  Map,
  Model,
  Nullable,
  OneToMany,
  OneToOne,
  References,
  String,
} from '../';
import codegen from '../codegen';
import { enumeration } from '../codegen/enum';
import { enumeration as enumerationColumn } from '../codegen/column';
import { Enum, Key } from '../public/fields/enums';
import * as Types from '../types';
import schema from './schema';

const generate = (
  schema: Types.Config['schema'],
  overrides: Partial<Types.Config> = {},
) =>
  codegen({
    schema,
    datasource: {
      url: 'env("DATABASE_URL")',
      provider: 'mysql',
      shadowDatabaseUrl: 'env("DATABASE_SHADOW_URL")',
      referentialIntegrity: 'prisma',
    },
    generators: [
      {
        name: 'client',
        provider: 'prisma-client-js',
        binaryTargets: [
          'native',
          'rhel-openssl-1.0.x',
          'linux-arm64-openssl-1.0.x',
          'darwin-arm64',
        ],
        previewFeatures: ['referentialIntegrity'],
      },
    ],
    ...overrides,
  });

describe('refract', () => {
  it('should generate the schema', () => {
    const { schema: prisma } = generate(schema);

    console.log(prisma);

    expect(replaceGeneratedTime(prisma)).toMatchSnapshot();
  });

  describe('should generate implicit many-to-many schema', () => {
    const Post = Model('Post');
    const Category = Model('Category');

    Post.Field('id', Int(Id, Default('autoincrement()'))).Relation(
      'categories',
      OneToMany(Category),
    );

    Category.Field('id', Int(Id, Default('autoincrement()'))).Relation(
      'posts',
      OneToMany(Post),
    );

    const implicit = generate([Post, Category]);
    console.log(implicit.schema);
  });

  describe('enum block generation', () => {
    it('should generate an enum', () => {
      const e = enumeration(
        Enum(
          'Example',
          Key('Qux'),
          Key('Foo', 'This is a comment'),
          Key('Bar', Map('Baz')),
        ),
      );

      expect(e).toMatchSnapshot();
    });

    it('should generate an enum with a comment', () => {
      const e = enumeration(
        Enum(
          'Example',
          'An Enum with a comment',
          Key('Qux'),
          Key('Foo', 'This is a comment'),
          Key('Bar', Map('Baz')),
        ),
      );

      expect(e).toMatchSnapshot();
    });
  });

  describe('enum column generation', () => {
    const e = Enum('Example', Key('Foo'), Key('Bar'));

    const asColumn = (e: Types.Fields.Field<'Enum'>): Types.Column<'Enum'> => ({
      name: 'test',
      ...(e as any),
    });

    const nonNullableNoDefault = e();
    expect(enumerationColumn(asColumn(nonNullableNoDefault))).toMatchSnapshot();

    const nullableNoDefault = e(Nullable);
    expect(enumerationColumn(asColumn(nullableNoDefault))).toMatchSnapshot();

    const nonNullableDefault = e('Foo');
    expect(enumerationColumn(asColumn(nonNullableDefault))).toMatchSnapshot();

    const nullableDefault = e('Foo', Nullable);
    expect(enumerationColumn(asColumn(nullableDefault))).toMatchSnapshot();
  });

  describe('should validate the schema and throw error when', () => {
    it('the other side of the relation is missing when specifying relation name', () => {
      const User = Model('User');
      const Post = Model('Post');

      User.Field('id', Int()).Relation(
        'posts',
        OneToMany(Post, 'WrittenPosts'),
      );
      Post.Field('authorId', Int(Nullable)).Relation(
        'author',
        ManyToOne(User, Fields('authorId'), References('id'), Nullable),
      );

      expect(() => generate([User, Post])).toThrow(
        "RelationshipErr: The other side of the relation 'posts' with name 'WrittenPosts' don't exist in model 'User'",
      );
    });

    it('the model have an ambiguous self relation', () => {
      const User = Model('User');

      User.Relation('friend', OneToOne(User));

      expect(() => generate([User])).toThrow(
        "RelationshipErr: The model 'User' have an ambiguous self relation. The fields 'friend' and 'User' both refer to 'User'. If they are part of the same relation add the same relation name for them with RelationName(<name>) modifier",
      );
    });

    it('some of the specified column in the Fields modifier are missing in the model', () => {
      const User = Model('User');
      const Post = Model('Post');

      User.Field('id', Int());
      Post.Relation(
        'author',
        ManyToOne(User, Fields('authorId'), References('id'), Nullable),
      );

      expect(() => generate([User, Post])).toThrow(
        "RelationshipErr: Columns in 'fields' don't exist in model 'Post': 'authorId'",
      );
    });

    it('some of the specified column in the References modifier are missing in the other side model', () => {
      const User = Model('User');
      const Post = Model('Post');

      Post.Field('authorId', Int()).Relation(
        'author',
        ManyToOne(User, Fields('authorId'), References('id'), Nullable),
      );

      expect(() => generate([User, Post])).toThrow(
        "RelationshipErr: Referenced columns in 'references' don't exist in model 'User': 'id'",
      );
    });

    it('the number of columns specified in the Fields and References modifiers are not equal', () => {
      const User = Model('User');
      const Post = Model('Post');

      User.Field('id', Int());
      Post.Field('authorId', Int())
        .Field('secondId', Int())
        .Relation(
          'author',
          ManyToOne(
            User,
            Fields('authorId', 'secondId'),
            References('id'),
            Nullable,
          ),
        );

      expect(() => generate([User, Post])).toThrow(
        "RelationshipErr: You must specify the same number of fields in 'fields' and 'references' for relation 'author' in model 'Post'",
      );
    });

    it('the types of columns specified in the Fields and References modifiers does not match', () => {
      const User = Model('User');
      const Post = Model('Post');

      User.Field('id', Int()).Field('name', String());
      Post.Field('authorId', Int())
        .Field('authorName', Int())
        .Relation(
          'author',
          ManyToOne(
            User,
            Fields('authorId', 'authorName'),
            References('id', 'name'),
            Nullable,
          ),
        );

      expect(() => generate([User, Post])).toThrow(
        "RelationshipErr: The type of the field 'authorName' in the model 'Post' does not match the type of the referenced field 'name' in model 'User'",
      );
    });

    it('there is a one-to-one relation with no scalar and not nullable', () => {
      const User = Model('User');
      const Post = Model('Post');

      User.Field('id', Int()).Relation('pinned', OneToOne(Post));
      Post.Field('pinnedById', Int()).Relation(
        'pinnedBy',
        OneToOne(User, Fields('pinnedById'), References('id'), Nullable),
      );

      expect(() => generate([User, Post])).toThrow(
        "RelationshipErr: The side of the one-to-one relation without a relation scalar must be optional\n(Model 'Post', relation 'pinned')",
      );
    });

    it('there is a scalar array field and not using the correct datasource', () => {
      const User = Model('User');

      User.Field('id', Int(Array));

      expect(() => generate([User])).toThrow(
        'ModifierErr: Scalar lists are only supported when using PostgreSQL, CockroachDB or MongoDB.',
      );
    });

    it('there is a scalar optional array field', () => {
      const User = Model('User');

      User.Field('id', Int(Array, Nullable));

      expect(() =>
        generate([User], {
          datasource: { provider: 'postgresql', url: 'url' },
        }),
      ).toThrow(
        "ModifierErr: Field 'id' cannot be an array and optional in the same time",
      );
    });
  });
});

function replaceGeneratedTime(schema: string): string {
  return schema.replace(/generated in [\d.]* ms/, 'generated in x ms');
}
