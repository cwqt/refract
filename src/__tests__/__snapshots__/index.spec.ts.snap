// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"	test Example"`;

exports[` 2`] = `"	test Example?"`;

exports[` 3`] = `"	test Example @default(Foo)"`;

exports[` 4`] = `"	test Example? @default(Foo)"`;

exports[`refract enum block generation should generate an enum 1`] = `
"enum Example {
	Qux
	// This is a comment
	Foo
	Bar @map(\\"Baz\\")
}"
`;

exports[`refract enum block generation should generate an enum with a comment 1`] = `
"// An Enum with a comment
enum Example {
	Qux
	// This is a comment
	Foo
	Bar @map(\\"Baz\\")
}"
`;

exports[`refract should generate the schema 1`] = `
"// refract https://github.com/cwqt/refract - generated in x ms -----------------
// datasource ----------------------------------------------------------------------

datasource db {
	url                   = env(\\"DATABASE_URL\\")
	provider              = \\"mysql\\"
	shadowDatabaseUrl     = env(\\"DATABASE_SHADOW_URL\\")
	referentialIntegrity  = \\"prisma\\"
}

// generators ----------------------------------------------------------------------

generator client {
	provider         = \\"prisma-client-js\\"
	binaryTargets    = [\\"native\\",\\"rhel-openssl-1.0.x\\",\\"linux-arm64-openssl-1.0.x\\",\\"darwin-arm64\\"]
	previewFeatures  = [\\"referentialIntegrity\\"]
}

// enums ---------------------------------------------------------------------------

// This is the Role Enum
enum Role {
	// This is the admin role
	ADMIN @map(\\"admin\\")
	USER @map(\\"user\\")
	// This is the owner role
	OWNER @map(\\"owner\\")
}

enum Foo {
	Bar
	Baz
}

// types ---------------------------------------------------------------------------

type Profile {
	fullname  String
	bio       String
}

// models --------------------------------------------------------------------------

// This is the User model
model User {
	id               Int        @id @default(autoincrement()) @map(\\"_id\\") @db.Value('foo')
	email            String     @unique @db.VarChar(4)
	name             String?
	// The user model
	height           Float      @default(1.8)
	role             Role?      @default(USER)
	foo              Foo
	bar              Foo?
	profile          Profile?
	// Relations are cool
	posts            Post[]     @relation(\\"WrittenPosts\\")
	pinned           Post?      @relation(\\"PinnedPost\\")
	createdAt        DateTime   @default(now())
	updatedAt        DateTime   @updatedAt @db.Date(6)
	@@index([mixin, index])
}

model Post {
	id                 Int       @id @default(autoincrement()) @db.UnsignedSmallInt
	published          Boolean   @default(false)
	title              String
	authorId           Int?
	author             User?     @relation(\\"WrittenPosts\\", fields: [authorId], references: [id], onUpdate: Restrict, onDelete: SetNull)
	pinnedById         Int?
	pinnedBy           User?     @relation(\\"PinnedPost\\", fields: [pinnedById], references: [id])
	stars              Star[]
	createdAt          DateTime  @default(now())
	updatedAt          DateTime  @updatedAt @db.Date(6)
	@@index([mixin, index])
	@@map(\\"comments\\")
}

model Star {
	id                     Int                      @id @default(autoincrement())
	decimal                Decimal                  @db.Decimal(10, 20)
	postId                 Int?
	post                   Post                     @relation(fields: [postId], references: [id])
	createdAt              DateTime                 @default(now())
	updatedAt              DateTime                 @updatedAt @db.Date(6)
	@@index([mixin, index])
	location               Unsupported(\\"polygon\\")?
	@@unique([A, B], map: \\"_AB_unique\\")
	// Block level comments?
	@@index([wow], map: \\"_B_index\\")
	@@map(\\"Group\\")
	@@fulltext([location, decimal])
}"
`;
